### ISSUE TITLE
about keep alive

### ISSUE THREAD

**Issue by wangkaisdu** [openresty/lua-resty-redis#70]
hi,I try that again. forgive my poor english please.

I copy the example from https://github.com/openresty/lua-resty-redis ,but it don't keep-alive, and I paste the network Package in the end.

## This is the example: 
```lua
location /test {
            content_by_lua '
                local redis = require "resty.redis"
                local red = redis:new()

                red:set_timeout(1000) -- 1 sec

                -- or connect to a unix domain socket file listened
                -- by a redis server:
                --     local ok, err = red:connect("unix:/path/to/redis.sock")

                local ok, err = red:connect("127.0.0.1", 6379)
                if not ok then
                    ngx.say("failed to connect: ", err)
                    return
                end

                ok, err = red:set("dog", "an animal")
                if not ok then
                    ngx.say("failed to set dog: ", err)
                    return
                end

                ngx.say("set result: ", ok)

                local res, err = red:get("dog")
                if not res then
                    ngx.say("failed to get dog: ", err)
                    return
                end

                if res == ngx.null then
                    ngx.say("dog not found.")
                    return
                end

                ngx.say("dog: ", res)

                red:init_pipeline()
                red:set("cat", "Marry")
                red:set("horse", "Bob")
                red:get("cat")
                red:get("horse")
                local results, err = red:commit_pipeline()
                if not results then
                    ngx.say("failed to commit the pipelined requests: ", err)
                    return
                end

                for i, res in ipairs(results) do
                    if type(res) == "table" then
                        if res[1] == false then
                            ngx.say("failed to run command ", i, ": ", res[2])
                        else
                            -- process the table value
                        end
                    else
                        -- process the scalar value
                    end
                end

                -- put it into the connection pool of size 100,
                -- with 10 seconds max idle time
                local ok, err = red:set_keepalive(10000, 100)
                if not ok then
                    ngx.say("failed to set keepalive: ", err)
                    return
                end

                -- or just close the connection right away:
                -- local ok, err = red:close()
                -- if not ok then
                --     ngx.say("failed to close: ", err)
                --     return
                -- end
            ';
        }
```

## This is the result:  

the client send fin package immediately after receive the data.

```
11:04:29.128486 IP localhost.59740 > localhost.6379: Flags [S], seq 3274653696, win 65535, options [mss 16344,nop,wscale 5,nop,nop,TS val 595791539 ecr 0,sackOK,eol], length 0
11:04:29.128529 IP localhost.6379 > localhost.59740: Flags [S.], seq 1325370336, ack 3274653697, win 65535, options [mss 16344,nop,wscale 5,nop,nop,TS val 595791539 ecr 595791539,sackOK,eol], length 0
11:04:29.128538 IP localhost.59740 > localhost.6379: Flags [.], ack 1, win 12759, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.128545 IP localhost.6379 > localhost.59740: Flags [.], ack 1, win 12759, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.128592 IP localhost.59740 > localhost.6379: Flags [P.], seq 1:38, ack 1, win 12759, options [nop,nop,TS val 595791539 ecr 595791539], length 37
11:04:29.128613 IP localhost.6379 > localhost.59740: Flags [.], ack 38, win 12758, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.128645 IP localhost.6379 > localhost.59740: Flags [P.], seq 1:6, ack 38, win 12758, options [nop,nop,TS val 595791539 ecr 595791539], length 5
11:04:29.128661 IP localhost.59740 > localhost.6379: Flags [.], ack 6, win 12759, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.128710 IP localhost.59740 > localhost.6379: Flags [P.], seq 38:60, ack 6, win 12759, options [nop,nop,TS val 595791539 ecr 595791539], length 22
11:04:29.128728 IP localhost.6379 > localhost.59740: Flags [.], ack 60, win 12757, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.128758 IP localhost.6379 > localhost.59740: Flags [P.], seq 6:21, ack 60, win 12757, options [nop,nop,TS val 595791539 ecr 595791539], length 15
11:04:29.128775 IP localhost.59740 > localhost.6379: Flags [.], ack 21, win 12758, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.128828 IP localhost.59740 > localhost.6379: Flags [P.], seq 60:172, ack 21, win 12758, options [nop,nop,TS val 595791539 ecr 595791539], length 112
11:04:29.128848 IP localhost.6379 > localhost.59740: Flags [.], ack 172, win 12754, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.128899 IP localhost.6379 > localhost.59740: Flags [P.], seq 21:51, ack 172, win 12754, options [nop,nop,TS val 595791539 ecr 595791539], length 30
11:04:29.128911 IP localhost.59740 > localhost.6379: Flags [.], ack 51, win 12757, options [nop,nop,TS val 595791539 ecr 595791539], length 0

11:04:29.128998 IP localhost.59740 > localhost.6379: Flags [F.], seq 172, ack 51, win 12757, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.129015 IP localhost.6379 > localhost.59740: Flags [.], ack 173, win 12754, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.129020 IP localhost.59740 > localhost.6379: Flags [.], ack 51, win 12757, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.129035 IP localhost.6379 > localhost.59740: Flags [F.], seq 51, ack 173, win 12754, options [nop,nop,TS val 595791539 ecr 595791539], length 0
11:04:29.129048 IP localhost.59740 > localhost.6379: Flags [.], ack 52, win 12757, options [nop,nop,TS val 595791539 ecr 595791539], length 0
```

**Comment by agentzh** [comment]
@wangkaisdu Change those `ngx.say` calls for error outputting in the code to `ngx.log(ngx.ERR, ...)` calls so that you can see if there's any errors before reaching the `set_keepalive` call in your nginx's `error.log` file. My hunch is that your Lua code aborts abnormally due to errors.

**Comment by agentzh** [comment]
@wangkaisdu Also, please note that, the maximum idle time in the connection pool configured in that example is 10 seconds. So connections are closed automatically by the client when they are idle in the pool for more than 10 sec.

**Comment by agentzh** [comment]
@wangkaisdu I've just tested that code sample on my side with the latest OpenResty and the connection pool works as expected:

```console
$ curl localhost:1984/test
set result: OK
dog: an animal
done

$ netstat -nt|grep 6379
tcp        0      0 127.0.0.1:43705         127.0.0.1:6379          ESTABLISHED
tcp        0      0 127.0.0.1:6379          127.0.0.1:43705         ESTABLISHED

$ curl localhost:1984/test
set result: OK
dog: an animal
done

$ netstat -nt|grep 6379
tcp        0      0 127.0.0.1:43705         127.0.0.1:6379          ESTABLISHED
tcp        0      0 127.0.0.1:6379          127.0.0.1:43705         ESTABLISHED

$ netstat -nt|grep 6379
tcp        0      0 127.0.0.1:43705         127.0.0.1:6379          ESTABLISHED
tcp        0      0 127.0.0.1:6379          127.0.0.1:43705         ESTABLISHED

$ curl localhost:1984/test
set result: OK
dog: an animal
done

$ netstat -nt|grep 6379
tcp        0      0 127.0.0.1:43705         127.0.0.1:6379          ESTABLISHED
tcp        0      0 127.0.0.1:6379          127.0.0.1:43705         ESTABLISHED

$ netstat -nt|grep 6379
tcp        0      0 127.0.0.1:43705         127.0.0.1:6379          ESTABLISHED
tcp        0      0 127.0.0.1:6379          127.0.0.1:43705         ESTABLISHED

$ curl localhost:1984/test
set result: OK
dog: an animal
done

$ netstat -nt|grep 6379
tcp        0      0 127.0.0.1:43705         127.0.0.1:6379          ESTABLISHED
tcp        0      0 127.0.0.1:6379          127.0.0.1:43705         ESTABLISHED
```

BTW, connection pools are per-worker. So in the test above, I configured only a single worker so as to ensure that my client requests always hit the same worker, and thus the same connection pool.

Also, I change the 10 sec max idle time to 3600 sec so that we can ensure the keep-alive connections are always there even when I am slow on typing :)

**Comment by agentzh** [comment]
@wangkaisdu I don't know how you initiated requests. If you initiate too many client connections that exceed the size limit of the connection pool, then ngx_lua will close the least-recently used idle connection in the pool and reuse the latest successfully used connection. Please refer to the documentation for more details. Well, I hope you can be more specific in your testing procedure so that I don't have to guess and waste both of our time :)

**Comment by wangkaisdu** [comment]
I close lua_code_cache for debug so the keep-alive can't work, when i open it the keep-alive can work well.

Thanks very much for you help , more important you method for your method to solve  problems.

**Comment by AlexProfi** [comment]
Hello
I get similar errors in nginx log
from my code

local redis = require "resty.redis"
local red = redis:new()

red:set_timeout(20)
--Connect to REDIS
local ok, err = red:connect(os.getenv("REDIS_SVC_SERVICE_HOST"), os.getenv("REDIS_PORT"))

if not ok then
    ngx.log(ngx.WARN, "Failed to connect to REDIS server: ", err)
end

local res_fingerprint, err = red:exists("test")

local ok, err = red:set_keepalive(0, 4096)
if not ok then
    ngx.log(ngx.WARN, "Failed to set keepalive: ", err)
end

How to resolve this



```
2019/01/22 18:52:06 [error] 17#17: *502179 lua tcp socket read timed out,        
client: 10.244.5.0, server: meteotravel.ru, request: "GET / HTTP/1.0", host:     
"meteotravel.ru"
2019/01/22 18:52:06 [warn] 17#17: *502179 [lua] access.lua:34: Failed to set     
keepalive: closed, client: 10.244.5.0, server: meteotravel.ru, request: "GET /   
HTTP/1.0", host: "meteotravel.ru"

```

**Comment by agentzh** [comment]
@AlexProfi Because your timeout threshold is too small (20 millisecond). A timed out redis connection cannot be reused anyway. Please do not hijack closed issues for unrelated problems. That's annoying. Thanks for your cooperation.
